src/coreclr/jit/valuenum.cpp

```diff
@@ -2134,6 +2134,35 @@ ValueNum ValueNumStore::VNForFunc(var_types typ, VNFunc func, ValueNum arg0VN, V
         return EvalFuncForConstantArgs(typ, func, arg0VN, arg1VN);
     }
 
+    if ((arg0VN == arg1VN) && !varTypeIsFloating(TypeOfVN(arg0VN)))
```

Comment:
I would expect such logic to live in `EvalUsingMathIdentity`.

I think it already does (some of) this.

Reply:
Aha, yep, majority of the improvements I was getting was from the wrong float folding :)
Looks like `EvalUsingMathIdentity` does the most important parts of this and the extra identities only result in a small number of improvements in non-test collections, so doesn't seem worth doing.

Reply:
Still I am pretty sure there are optimizations in morph which we can do here as well 🙂

Reply:
```
cat morph.cpp >> valuenum.cpp
```